[docs] performance advisor for YBA (#17070)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-platform/alerts-monitoring/performance-advisor.md

**Fix the problem**

If you see query load skew, contact {{% support-platform %}}.

### CPU skew and CPU usage

Advisor monitors CPU use to determine whether any nodes become hot spots. Advisor flags nodes where CPU use on a node is 50% greater than the other nodes in the cluster in the past hour (skew), or CPU use for a node exceeds 80% for 10 minutes (usage). CPU skew recommendation is not raised until node has at least 50% average CPU usage in the last 10 minutes.

CPU skew and high usage can be caused by a number of issues:

- A node has a higher number of connections.
- A node is receiving a higher number of queries.
- Set of rows or specific rows are accessed frequently by queries executing across different nodes in the cluster.

Any of these conditions can result in a hot spot.

**Fix the problem**

Review the sharding strategies for your primary and secondary indexes. Consistent hash sharding is better for scalability and preventing hot spots, while range sharding is better for range-based queries.

> ### Rejected connections

Reviewer: add recommendation to the header

Advisor monitors rejected YSQL connections. Advisor flags nodes, which rejected 1 or more connections in the last hour. By default each YB-TServer can handle up to 300 simultaneous connections. This number can be configured with [ysql-max-connections](../../../reference/configuration/yb-tserver/#ysql-max-connections) YB-Tserver flag.

Rejected connections can be caused by:

- Connection skew - all clients are connected to one YB-TServer.
- Too many clients or too large connection pool sizes.
- Connection pooling not used on the client side and too many simultaneous requests.
- Load balancer configured incorrectly.

**Fix the problem**

> Review your YSQL clients configuration to see, if connection pooling is configured and pool sizes are correct. If you need more connections, review YB-TServer memory usage. If enough free memory is present, increase the number of `ysql-max-connections`. Otherwise increase the number of YB-TServers to spread connections across them.

Reviewer 2: ```suggestion
Review your YSQL client's configuration to see if connection pooling is configured and pool sizes are correct. If you need more connections, review YB-TServer memory usage. If enough free memory is present, increase the number of `ysql-max-connections`. Otherwise increase the number of YB-TServers to spread connections across them.
```

### Hot shard

> Advisor monitors per-table read or write statistics on each node. Advisor flags nodes, which processed 8x more reads or writes for a particular table in the last 10 minutes. Hot shard recommendation is not raised until a node processes at least 600 reads or writes during the last 10 minutes.

Reviewer 2: ```suggestion
Advisor monitors per-table read and write statistics on each node. Advisor flags nodes that processed 8x more reads or writes for a particular table in the last 10 minutes. Hot shard recommendations are not raised until a node processes at least 600 reads or writes during the last 10 minutes.
```

> Advisor monitors per-table read and write statistics on each node. Advisor flags nodes that processed 8x more reads or writes for a particular table in the last 10 minutes. Hot shard recommendations are not raised until a node processes at least 600 reads or writes during the last 10 minutes.

Reviewer: In YugabyteDB, a "hot shard" problem can occur when one or a few tablets (or shards) of a table receive a disproportionately high volume of read or write requests, compared to the other shards. This can lead to performance issues, as the hot shards become a bottleneck and are unable to handle the load efficiently.

When a hot shard problem occurs, it can cause uneven data distribution across the cluster, leading to slower query response times and increased resource utilization on the hot shards. In extreme cases, it can lead to node failures or cluster instability.

> Hot shard can be caused by:

Reviewer 2: ```suggestion
A hot shard can be caused by:
```

- Using range sharding with sub-optimal shard configuration.
- An application reading or writing one particular row too often.

**Fix the problem**

> If range sharding is used, analyze reads or writes and recreate the table with more fair shards distribution across nodes.

Reviewer 2: ```suggestion
If range sharding is used, analyze reads or writes and recreate the table with a more balanced shard distribution across nodes.
```

> If range sharding is used, analyze reads or writes and recreate the table with more fair shards distribution across nodes.

Reviewer 2: How?

Reviewer 2: But also, mechanically, how do you change the shard distribution - are there commands we can point to, or anything like that

## Limitations

- Currently, you can't schedule periodic performance advisor runs.

> - Currently, you can configure performance recommendation thresholds only via the API.

Reviewer 2: ```suggestion
- Currently, you can only configure performance recommendation thresholds via the API.
```

> - Performance recommendation can be marked resolved only via UI; this state is not persisted.

Reviewer 2: ```suggestion
- Performance recommendations can only be marked resolved via the UI; this state is not persisted.
```

- Limited set of performance recommendations.

> 

Reviewer: Are there any limitations on scale or performance of the feature when run at scale?

Reviewer 3: I do not believe there will be limitations with respect to performance of the feature when run at scale, but since we do manual runs and not periodic runs, it takes a longer to scan recommendations and display results.
Cc @anmalysh-yb

Reviewer 4: Currently the only limitation is that the scan for a universe with large number of nodes/databases can take more time.
We may have more limitations once we allow periodic runs though.
